crates/indexer/src/models/token_models/petra_activities.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::models::coin_models::coin_activities::CoinActivity;
+use crate::models::token_models::token_activities::TokenActivity;
+use crate::schema::petra_activities;
+use crate::util::standardize_address;
+use aptos_api_types::{Transaction as APITransaction, WriteSetChange as APIWriteSetChange};
+use field_count::FieldCount;
+use serde::{Deserialize, Serialize};
+use std::collections::HashSet;
+
+#[derive(Debug, Deserialize, FieldCount, Identifiable, Insertable, Serialize)]
+#[diesel(primary_key(account_address, transaction_version))]
+#[diesel(table_name = petra_activities)]
+pub struct PetraActivity {
+    pub transaction_version: i64,
+    pub account_address: String,
+    pub coin_activities: serde_json::Value,
+    pub token_activities: serde_json::Value,
+}
+
+const EMPTY_JSON_ARRAY: serde_json::Value = serde_json::Value::Array(vec![]);
+
+impl PetraActivity {
+    pub fn from_transaction(
+        transaction: &APITransaction,
+        coin_activities: Vec<CoinActivity>,
+        token_activities: Vec<TokenActivity>,
+    ) -> Vec<Self> {
+        let mut petra_activities = vec![];
+
+        if let APITransaction::UserTransaction(user_txn) = transaction {
+            let version = user_txn.info.version.0 as i64;
+            let mut addresses = HashSet::new();
+            let coin_activities_json =
+                serde_json::to_value(coin_activities).unwrap_or(EMPTY_JSON_ARRAY);
+            let token_activities_json =
+                serde_json::to_value(token_activities).unwrap_or(EMPTY_JSON_ARRAY);
+
+            // Add the transaction sender.
+            addresses.insert(user_txn.request.sender.to_string());
+
+            // Add any other accounts referenced in the writeset.
+            for wsc in user_txn.info.changes.iter() {
```

Comment:
i don't think you need this given all the content are coming from events? You should be able to get this from reusing the TokenActivity and CoinActivity objects.

Reply:
Good point, will update.